### ISSUE TITLE
Test failure of EB 2.4.0

### ISSUE THREAD

**Issue by jhein32** [hpcugent/easybuild-framework#1478]
Hi,

I am doing the following:

```
export PREFIX=/home/hein/EasyBuild_Sandbox/easyTest
mkdir $PREFIX
curl -O https://raw.githubusercontent.com/hpcugent/easybuild-framework/develop/easybuild/scripts/bootstrap_eb.py

bootstrap
python bootstrap_eb.py $PREFIX

# setting up env
export MODULEPATH=/home/hein/EasyBuild_Sandbox/easyTest/modules/all:$MODULEPATH
export EASYBUILD_MODULES_TOOL=Lmod
export EASYBUILD_INSTALLPATH=/home/hein/EasyBuild_Sandbox/easyTest

# testing

export TEST_EASYBUILD_MODULES_TOOL=Lmod
python -m test.framework.suite
```

And get the following:

```
WARNING: xmlrunner module not available, falling back to using unittest...

.......................Skipping test_from_pr, no GitHub token available?
.Skipping test_from_pr, no GitHub token available?
.F..............................(skipping GitRepository test)
.(skipping HgRepository test)
..(skipping SvnRepository test)
..............................Skipping test_dep_graph, since pygraph is not available
....Skipping test_dump_autopep8, since autopep8 is not available
........................................................................................................Skipping test_from_pr, no GitHub token available?
............................Skipping test_fetch_easyconfigs_from_pr, no GitHub token available?
..Skipping test_read, no GitHub token available?
.Skipping test_read_api, no GitHub token available?
.Skipping test_walk, no GitHub token available?
.............................................err: 'Build of /home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/easyconfigs/toy-0.0.eb failed (err: \'build failed (first 300 chars): Sanity check failed: loading fake module failed: \\\'Changing environment as dictated by module failed: name \\\\\\\'false\\\\\\\' is not defined (stdout: os.environ["LMOD_DEFAULT_MODULEPATH"] = "/tmp/eb-OzaBOs/eb-QtlW46/eb-V4OSzZ/eb-xtSjgQ/eb-CtwkE8/eb-Bbufea/eb-8_ynb7/eb-bVSTte/tmpd0ohIq/modules/all:/user/aurora\')'
F.....F..F...FF...............................................GC3Pie not available, skipping test
................Skipping test_join (no PyYAML available)
.Skipping test_parse_yeb (no PyYAML available)
.
======================================================================
FAIL: test_generate_cmd_line (test.framework.options.CommandLineOptionsTest)
Test for generate_cmd_line.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/options.py", line 1674, in test_generate_cmd_line
    self.assertEqual(ebopts.generate_cmd_line(), [])
AssertionError: Lists differ: ['--buildpath=/local/easybuild... != []

First list contains 4 additional elements.
First extra element 0:
--buildpath=/local/easybuild/build

+ []
- ['--buildpath=/local/easybuild/build',
-  '--installpath=/sw/easybuild',
-  '--modules-tool=Lmod',
-  '--packagepath=/sw/easybuild/packages']

======================================================================
FAIL: test_toy_hidden (test.framework.toy_build.ToyBuildTest)
Test installing a hidden module.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/toy_build.py", line 734, in test_toy_hidden
    self.assertTrue(os.path.exists(toy_module), 'Found hidden module %s' % toy_module)
AssertionError: Found hidden module /tmp/eb-OzaBOs/eb-QtlW46/eb-V4OSzZ/eb-xtSjgQ/eb-CtwkE8/eb-Bbufea/eb-8_ynb7/eb-bVSTte/tmpd0ohIq/modules/all/toy/.0.0

======================================================================
FAIL: test_XDG_CONFIG_env_vars (test.framework.config.EasyBuildConfigTest)
Test effect of XDG_CONFIG* environment variables on default configuration.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/config.py", line 433, in test_XDG_CONFIG_env_vars
    self.assertEqual(eb_go.options.configfiles, cfg_files)
AssertionError: Lists differ: ['/etc/easybuild.d/config.cfg'... != ['/tmp/eb-OzaBOs/eb-QtlW46/eb-...

First differing element 0:
/etc/easybuild.d/config.cfg
/tmp/eb-OzaBOs/eb-QtlW46/eb-V4OSzZ/eb-xtSjgQ/eb-CtwkE8/eb-Bbufea/eb-8_ynb7/eb-Nrb9bl/homedir/.config/easybuild/config.cfg

First list contains 1 additional elements.
First extra element 1:
/tmp/eb-OzaBOs/eb-QtlW46/eb-V4OSzZ/eb-xtSjgQ/eb-CtwkE8/eb-Bbufea/eb-8_ynb7/eb-Nrb9bl/homedir/.config/easybuild/config.cfg

- ['/etc/easybuild.d/config.cfg',
-  '/tmp/eb-OzaBOs/eb-QtlW46/eb-V4OSzZ/eb-xtSjgQ/eb-CtwkE8/eb-Bbufea/eb-8_ynb7/eb-Nrb9bl/homedir/.config/easybuild/config.cfg']
? ^

+ ['/tmp/eb-OzaBOs/eb-QtlW46/eb-V4OSzZ/eb-xtSjgQ/eb-CtwkE8/eb-Bbufea/eb-8_ynb7/eb-Nrb9bl/homedir/.config/easybuild/config.cfg']
? ^


======================================================================
FAIL: test_default_config (test.framework.config.EasyBuildConfigTest)
Test default configuration.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/config.py", line 118, in test_default_config
    self.assertEqual(config_options['buildpath'], os.path.join(eb_homedir, 'build'))
AssertionError: '/local/easybuild/build' != '/home/hein/.local/easybuild/build'

======================================================================
FAIL: test_generaloption_config (test.framework.config.EasyBuildConfigTest)
Test new-style configuration (based on generaloption).
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/config.py", line 141, in test_generaloption_config
    self.assertEqual(install_path(), os.path.join(prefix, 'software'))
AssertionError: '/sw/easybuild/software' != '/tmp/eb-OzaBOs/eb-QtlW46/eb-V4OSzZ/eb-xtSjgQ/eb-CtwkE8/eb-Bbufea/eb-8_ynb7/eb-goD8Dl/eb-DQr9gD/tmphvb5TS/testprefix/software'

======================================================================
FAIL: test_generaloption_config_file (test.framework.config.EasyBuildConfigTest)
Test use of new-style configuration file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/config.py", line 312, in test_generaloption_config_file
    self.assertEqual(install_path(), os.path.join(topdir, 'software'))  # default
AssertionError: '/sw/easybuild/software' != '/home/hein/.local/easybuild/software'

----------------------------------------------------------------------
Ran 354 tests in 2128.644s

FAILED (failures=6)
ERROR: Not all tests were successful.
Log available at /tmp/eb-OzaBOs/eb-QtlW46/easybuild-tests-IUH6Au.log 
```

It doesn't allow me to attach a gz-version of the log file :(  Please let me know how to submit this.

Best wishes
  Joachim

**Comment by boegel** [comment]
@jhein32: this seems like a duplicate of #1467, i.e. the tests picking up your configuration file from the default location

Can you confirm that this is the case, i.e. can you try to temporarily rename your configuration file, and rerun the tests?

We'll figure out a way to make the tests ignore the configuration file at the default location.

**Comment by jhein32** [comment]
@boegel: Could you give me pointers on where this config file is to be found?  (I am new to easybuild and still trying to get my head around it).  I don't seem to have one in $HOME/.config/ which is the location pointed out in #1467.

**Comment by boegel** [comment]
@jhein32: you can check with `eb --show-default-configfiles`, but if you're not aware of the location of the default config file, you probably didn't specify anything in it?

How did you configure EasyBuild to use `/sw/easybuild/software` as a software installation prefix?

**Comment by jhein32** [comment]
@boegel 

As we discussed earlier, the cluster has an "experimental" install of Easybuild 2.3.0 for all users (done by a colleague, who is currently at a conference, so not so easy to get hold off).  

I am a bit surprised how much that creeps into the system.  I found a config in `/etc`, which I renamed.  Now I am done to 1 failure.  Output is as follows:

    WARNING: xmlrunner module not available, falling back to using unittest...

    .......................Skipping test_from_pr, no GitHub token available?
    .Skipping test_from_pr, no GitHub token available?
    ................................(skipping GitRepository test)
    .(skipping HgRepository test)
    ..(skipping SvnRepository test)
    ..............................Skipping test_dep_graph, since pygraph is not available
    ....Skipping test_dump_autopep8, since autopep8 is not available
    ........................................................................................................Skipping test_from_pr, no GitHub token available?
    ............................Skipping test_fetch_easyconfigs_from_pr, no GitHub token available?
    ..Skipping test_read, no GitHub token available?
    .Skipping test_read_api, no GitHub token available?
    .Skipping test_walk, no GitHub token available?
    .............................................err: 'Build of /home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/easyconfigs/toy-0.0.eb failed (err: \'build failed (first 300 chars): Sanity check failed: loading fake module failed: \\\'Changing environment as dictated by module failed: name \\\\\\\'false\\\\\\\' is not defined (stdout: os.environ["LMOD_DEFAULT_MODULEPATH"] = "/tmp/eb-tPMvh_/eb-LyWYDd/eb-mRkRkE/eb-7NAitR/eb-zfYxHJ/eb-POG4ov/eb-VZ1AlO/eb-C5O6jU/tmp_vG4Bt/modules/all:/user/aurora\')'
    F.............................................................GC3Pie not available, skipping test
    ................Skipping test_join (no PyYAML available)
    .Skipping test_parse_yeb (no PyYAML available)
    .
    ======================================================================
    FAIL: test_toy_hidden (test.framework.toy_build.ToyBuildTest)
    Test installing a hidden module.
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/hein/EasyBuild_Sandbox/easyTest/software/EasyBuild/2.4.0/lib/python2.7/site-packages/easybuild_framework-2.4.0-py2.7.egg/test/framework/toy_build.py", line 734, in test_toy_hidden
    self.assertTrue(os.path.exists(toy_module), 'Found hidden module %s' % toy_module)
    AssertionError: Found hidden module /tmp/eb-tPMvh_/eb-LyWYDd/eb-mRkRkE/eb-7NAitR/eb-zfYxHJ/eb-POG4ov/eb-VZ1AlO/eb-C5O6jU/tmp_vG4Bt/modules/all/toy/.0.0

    ----------------------------------------------------------------------
    Ran 354 tests in 2116.636s

    FAILED (failures=1)
    ERROR: Not all tests were successful.
    Log available at /tmp/eb-tPMvh_/eb-LyWYDd/easybuild-tests-aF0S7L.log

**Comment by boegel** [comment]
@jhein32: I'm assuming we're talking a configuration file that matches `/etc/easybuild.d/*.cfg`, i.e. the default system location where EB picks up configuration files?
Yes, the tests are not robust yet against having that in place.

I know @geimer ran into the exact same failing tests while he was working on a change in the framework (I think it was with #1271).

Something is going wrong when installing a hidden module, in the sense that the module file is never generated, while it should be.

Can you try changing the call to `self.test_toy_build(...)` in the `test_toy_hidden` method to `print self.test_toy_build(...)` in `test/framework/toy_build.py` and rerunning only those tests using `python -m test.framework.toy_build` (just to speed things up), and provide us the output?

**Comment by jhein32** [comment]
Hi,
You got a tar ball by ordinary mail.
Joachim

**Comment by boegel** [comment]
@jhein32: you're hitting a known (and fixed) bug in Lmod v6.0.1, which manifests itself as follows:

```
stderr: /bin/lua: /sw/lmod/lmod/libexec/MT.lua:565: attempt to compare nil with string
```

@rtmclay hasn't uploaded a new official release yet via SourceForge, but you can try with the latest tagged version from GitHub: https://github.com/TACC/Lmod/releases

**Comment by jhein32** [comment]
@boegel 
Ok, I tried with lmod 6.0.24 and hiding the config files.  The tests now pass without any failure :)  Thanks for your help and support.
I assume the hidding of config files will not be required in the future release.

**Comment by boegel** [comment]
@jhein32: hiding the configuration file is only needed to avoid that the tests pick up on it; there's a dedicated issue for this (#1467), and the plan is indeed to fix this (not sure when that will happen though)

Since we have things figured out here, I propose we close this issue?

**Comment by boegel** [comment]
fixed in #1513
